examples/inference/distributed/florence2.py

```diff
@@ -0,0 +1,203 @@
+    distributed_state = PartialState()
+
+    if distributed_state.is_main_process:
+        output_dir.mkdir(exist_ok=True)
+
+    model = AutoModelForCausalLM.from_pretrained(
+        model_name,
+        device_map=distributed_state.device,
+        torch_dtype=torch.float16,
+        trust_remote_code=True,
+    )
+
+    processor = AutoProcessor.from_pretrained(model_name, trust_remote_code=True, clean_up_tokenization_spaces=True)
+
+    class ExistsFilter:
+        def __init__(self, output_dir: Union[pathlib.Path, str]):
+            current_training_img_hashes = [f.split(".jpg")[0] for f in os.listdir(output_dir) if f.endswith(".jpg")]
+            self.current_training_img_hashes = set(current_training_img_hashes)
+            if distributed_state.is_main_process:
+                print(f"Existing images found: {len(self.current_training_img_hashes)}.")
+
+        def __call__(self, x):
+            if len(self.current_training_img_hashes) > 0:
+                if x["img_hash"] in self.current_training_img_hashes:
+                    return False
+                else:
+                    return True
+            else:
+                return True
+
+    def pil_to_bytes(pil_image: Image.Image, format="PNG"):
+        byte_arr = io.BytesIO()
+        pil_image.save(byte_arr, format=format)
+        im_bytes = byte_arr.getvalue()
+        return im_bytes
+
+    def preprocess_fn(sample):
+        image: Image.Image = sample["jpg"].convert("RGB")
+        img_hash = insecure_hashlib.sha1(image.tobytes()).hexdigest()
+        img_bytes = pil_to_bytes(image)
+        return {
+            "img_bytes": img_bytes,
+            "img_hash": img_hash,
+            "original_caption": sample["txt"],
+            "height": image.height,
+            "width": image.width,
+        }
+
+    def collate_fn(examples, processor):
+        original_images = [Image.open(io.BytesIO(sample["img_bytes"])).convert("RGB") for sample in examples]
+        inputs = processor(
+            text=[prompt] * len(original_images),
+            images=original_images,
+            return_tensors="pt",
+        )
+
+        img_bytes = [example["img_bytes"] for example in examples]
+        captions = [example["original_caption"] for example in examples]
+        heights = [example["height"] for example in examples]
+        widths = [example["width"] for example in examples]
+        inputs.update(
+            {
+                "img_bytes": img_bytes,
+                "original_captions": captions,
+                "heights": heights,
+                "widths": widths,
+            }
+        )
+        return dict(inputs)
+
+    dataset = (
+        wds.WebDataset(
+            data_path,
+            handler=wds.warn_and_continue,
+            nodesplitter=None,
+            shardshuffle=False,
+        )
+        .decode("pil", handler=wds.warn_and_continue)
+        .map(preprocess_fn, handler=wds.warn_and_continue)
+        .select(ExistsFilter(output_dir))
```

Comment:
Should this be applied only when `len(current_training_img_hashes) > 0`?

Reply:
```python
def __call__(self, x):
    if len(self.current_training_img_hashes) > 0:
        if x["img_hash"] in self.current_training_img_hashes:
            return False
        else:
            return True
    else:
        return True
```
Can do, shouldn't make much difference though, and it's a little awkward due to the way [`select`](https://github.com/webdataset/webdataset/blob/fe2ac2b39635093ed9bed70ee31b435fcd2dbb71/webdataset/compat.py#L161-L173) [works](https://github.com/webdataset/webdataset/blob/fe2ac2b39635093ed9bed70ee31b435fcd2dbb71/webdataset/filters.py#L418-L431), we'd have to do something like
```python
exist_filter = ExistsFilter(output_dir)

dataset = (
    wds.WebDataset(
        data_path,
        handler=wds.warn_and_continue,
        nodesplitter=None,
        shardshuffle=False,
    )
    .decode("pil", handler=wds.warn_and_continue)
    .map(preprocess_fn, handler=wds.warn_and_continue)
)
if len(exist_filter.current_training_img_hashes) > 0:
    dataset = dataset.select(ExistsFilter(output_dir))
dataset = dataset.batched(
    batch_size,
    partial=False,
    collation_fn=partial(collate_fn, processor=processor),
)
```
WDYT?

Reply:
Yes, that is exactly what we will have to do.

Reply:
Done.